### ISSUE TITLE
SIGSEGV / Core Dump when deleting buffer containing a terminal

### ISSUE THREAD

**Issue by brettanomyces** [neovim/neovim#3112]
Using the following script which I call `nvimex` to send "bdelete!" command to Neovim results in a SIGSEGV or a "Aborted (Core Dumped)" the second time it is run.

```
#!/usr/bin/python2
from __future__ import print_function
from docopt import docopt
from neovim import attach
from neovim.api import NvimError
import os
import sys
__doc__ = """nvim-command - Run a command in the running neovim session.
Usage:
    nvim-command <command>..."""
if __name__ == "__main__":
    nvim = None
    try:
        nvim = attach("socket", path=os.environ["NVIM_LISTEN_ADDRESS"])
    except KeyError:
        print("No neovim session found.", file=sys.stderr)
        sys.exit(1)
    try:
        nvim.command(" ".join(docopt(__doc__)["<command>"]))
    except NvimError as e:
        print(e, file=sys.stderr)
        sys.exit(1)
```

To reliably reproduce (on my machine at least)
``` 
nvim -U NONE
:term
nvimex "bd!"
:term
nvimex "bd!"
Aborted (Core Dumped)
```

```
yukichb@yukichb-fm:~$ nvim --version
NVIM 0.0.0-alpha+201507261542 (compiled Jul 27 2015 09:13:22)
Commit: a497bf4bfdc79b6fa52a846d7d0a884ba007f452
Build type: Debug
Compilation: /usr/bin/cc -Wconversion -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -DINCLUDE_GENERATED_DECLARATIONS -DHAVE_CONFIG_H -I/home/yukichb/Workspace/neovim/build/config -I/home/yukichb/Workspace/neovim/src -I/home/yukichb/Workspace/neovim/.deps/usr/include -I/home/yukichb/Workspace/neovim/.deps/usr/include -I/home/yukichb/Workspace/neovim/.deps/usr/include/luajit-2.0 -I/home/yukichb/Workspace/neovim/.deps/usr/include -I/home/yukichb/Workspace/neovim/.deps/usr/include -I/home/yukichb/Workspace/neovim/.deps/usr/include -I/home/yukichb/Workspace/neovim/.deps/usr/include -I/usr/include -I/home/yukichb/Workspace/neovim/build/src/nvim/auto -I/home/yukichb/Workspace/neovim/build/include
Compiled by yukichb@yukichb-fm

Optional features included (+) or not (-): +acl   +iconv    +jemalloc 
For differences from Vim, see :help vim-differences

   system vimrc file: "$VIM/nvimrc"
     user vimrc file: "~/.nvimrc"
 2nd user vimrc file: "~/.nvim/nvimrc"
      user exrc file: "~/.exrc"
  fall-back for $VIM: "/usr/local/share/nvim"
```

**Comment by brettanomyces** [comment]
tested with both fish and bash as the shell

**Comment by brettanomyces** [comment]
The message I get when a SIGSEGV occurs (as opposed to a core dump) is:

    fish: "nvim cd.fish" terminated by signal SIGSEGV (Address boundary error)

**Comment by brettanomyces** [comment]
Appears to have been resolved by b732a27c4e568ff0d56c74348e1c3fdf38364b81

**Comment by oni-link** [comment]
With `test.py`:
```python
#!/usr/bin/env python

from neovim import attach
import os

nvim = attach("socket", path=os.environ["NVIM_LISTEN_ADDRESS"])
nvim.command(" bd!")
```

and `nvim -u NONE -c "term python test.py"` I still get

```c
=================================================================
==23058==ERROR: AddressSanitizer: heap-use-after-free on address 0x625000004958 at pc 0x00000133ec6a bp 0x7ffe64dc1480 sp 0x7ffe64dc1478
READ of size 8 at 0x625000004958 thread T0
    #0 0x133ec69 in terminal_enter /home/oni-link/git/neovim/src/nvim/terminal.c:355:16
    #1 0x9728f1 in edit /home/oni-link/git/neovim/src/nvim/edit.c:257:5
    #2 0xed7df6 in normal_cmd /home/oni-link/git/neovim/src/nvim/normal.c:1056:13
    #3 0xea8325 in main_loop /home/oni-link/git/neovim/src/nvim/main.c:747:7
    #4 0xe93ed8 in main /home/oni-link/git/neovim/src/nvim/main.c:534:3
    #5 0x7f84e72b7f9f in __libc_start_main (/lib64/libc.so.6+0x1ff9f)
    #6 0x4420b5 in _start (/home/oni-link/git/neovim/build/bin/nvim+0x4420b5)

0x625000004958 is located 8280 bytes inside of 9344-byte region [0x625000002900,0x625000004d80)
freed by thread T0 here:
    #0 0x4d7fe0 in __interceptor_cfree.localalias.1 /home/oni-link/git/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:35
    #1 0xac5a74 in xfree /home/oni-link/git/neovim/src/nvim/memory.c:108:3
    #2 0x1344806 in terminal_destroy /home/oni-link/git/neovim/src/nvim/terminal.c:436:3
    #3 0xe82d15 in term_close /home/oni-link/git/neovim/src/nvim/eval.c:21246:3
    #4 0x133c592 in terminal_close /home/oni-link/git/neovim/src/nvim/terminal.c:280:5
    #5 0xba7742 in close_buffer /home/oni-link/git/neovim/src/nvim/buffer.c:394:5
    #6 0xbbc2ea in empty_curbuf /home/oni-link/git/neovim/src/nvim/buffer.c:831:5
    #7 0xbb261b in do_buffer /home/oni-link/git/neovim/src/nvim/buffer.c:975:14
    #8 0xbb9cf7 in do_bufdel /home/oni-link/git/neovim/src/nvim/buffer.c:721:11
    #9 0x12873b9 in ex_bunload /home/oni-link/git/neovim/src/nvim/ex_docmd.c:4331:17
    #10 0x1238547 in do_one_cmd /home/oni-link/git/neovim/src/nvim/ex_docmd.c:2195:5
    #11 0x121730e in do_cmdline /home/oni-link/git/neovim/src/nvim/ex_docmd.c:621:20
    #12 0x121d5f5 in do_cmdline_cmd /home/oni-link/git/neovim/src/nvim/ex_docmd.c:293:10
    #13 0x1551ae7 in vim_command /home/oni-link/git/neovim/src/nvim/api/vim.c:44:3
    #14 0x51a827 in handle_vim_command /home/oni-link/git/neovim/build/src/nvim/auto/msgpack_dispatch.c:723:3
    #15 0x15c1cfd in on_request_event /home/oni-link/git/neovim/src/nvim/msgpack_rpc/channel.c:483:19
    #16 0x1615da5 in process_events_from /home/oni-link/git/neovim/src/nvim/event/loop.c:138:5
    #17 0x16172ca in loop_process_event /home/oni-link/git/neovim/src/nvim/event/loop.c:95:3
    #18 0x133ef14 in terminal_enter /home/oni-link/git/neovim/src/nvim/terminal.c:384:9
    #19 0x9728f1 in edit /home/oni-link/git/neovim/src/nvim/edit.c:257:5
    #20 0xed7df6 in normal_cmd /home/oni-link/git/neovim/src/nvim/normal.c:1056:13
    #21 0xea8325 in main_loop /home/oni-link/git/neovim/src/nvim/main.c:747:7
    #22 0xe93ed8 in main /home/oni-link/git/neovim/src/nvim/main.c:534:3
    #23 0x7f84e72b7f9f in __libc_start_main (/lib64/libc.so.6+0x1ff9f)

previously allocated by thread T0 here:
    #0 0x4d82f0 in calloc /home/oni-link/git/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:56
    #1 0xac5b12 in xcalloc /home/oni-link/git/neovim/src/nvim/memory.c:122:15
    #2 0x1336c50 in terminal_open /home/oni-link/git/neovim/src/nvim/terminal.c:190:18
    #3 0xe1a7c7 in f_termopen /home/oni-link/git/neovim/src/nvim/eval.c:16113:20
    #4 0xd0329d in call_func /home/oni-link/git/neovim/src/nvim/eval.c:7105:11
    #5 0xd17094 in get_func_tv /home/oni-link/git/neovim/src/nvim/eval.c:6954:11
    #6 0xd1153b in ex_call /home/oni-link/git/neovim/src/nvim/eval.c:2798:9
    #7 0x1238547 in do_one_cmd /home/oni-link/git/neovim/src/nvim/ex_docmd.c:2195:5
    #8 0x121730e in do_cmdline /home/oni-link/git/neovim/src/nvim/ex_docmd.c:621:20
    #9 0x121d5f5 in do_cmdline_cmd /home/oni-link/git/neovim/src/nvim/ex_docmd.c:293:10
    #10 0x12aef68 in ex_terminal /home/oni-link/git/neovim/src/nvim/ex_docmd.c:9408:3
    #11 0x1238547 in do_one_cmd /home/oni-link/git/neovim/src/nvim/ex_docmd.c:2195:5
    #12 0x121730e in do_cmdline /home/oni-link/git/neovim/src/nvim/ex_docmd.c:621:20
    #13 0x121d5f5 in do_cmdline_cmd /home/oni-link/git/neovim/src/nvim/ex_docmd.c:293:10
    #14 0xea6191 in exe_commands /home/oni-link/git/neovim/src/nvim/main.c:1781:5
    #15 0xe93915 in main /home/oni-link/git/neovim/src/nvim/main.c:497:5
    #16 0x7f84e72b7f9f in __libc_start_main (/lib64/libc.so.6+0x1ff9f)

SUMMARY: AddressSanitizer: heap-use-after-free /home/oni-link/git/neovim/src/nvim/terminal.c:355:16 in terminal_enter
```

In the first loop iteration of `terminal_enter()` the terminal is freed (at backtrace '#18'). When the loop checks it's condition for a second iteration, it uses already freed memory (at first backtrace '#0').
